torchgeo/datasets/mmearth.py

```diff
@@ -0,0 +1,595 @@
+            bands: available bands for the modality
+
+        Returns:
+            processed data
+        """
+        # band selection for modality
+        indices = self._select_indices_for_modality(modality, bands)
+        data = data[indices, ...]
+
+        # See https://github.com/vishalned/MMEarth-train/blob/8d6114e8e3ccb5ca5d98858e742dac24350b64fd/mmearth_dataset.py#L69
+        if modality == 'dynamic_world':
+            # first replacs 0 with nan then assign new labels to have 0-index classes
+            data = np.where(data == self.no_data_vals[modality], np.nan, data)
+            old_values = [1, 2, 3, 4, 5, 6, 7, 8, 9, np.nan]
+            new_values = [0, 1, 2, 3, 4, 5, 6, 7, 8, np.nan]
+            for old, new in zip(old_values, new_values):
+                data = np.where(data == old, new, data)
+
+            # need to replace nan with a no-data value and get long tensor
+            # maybe also 255 like esa_worldcover
+            tensor = torch.from_numpy(data)
+
+        elif modality == 'esa_worldcover':
+            old_values = [10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 100, 255]
+            new_values = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 255]
+            for old, new in zip(old_values, new_values):
+                data = np.where(data == old, new, data)
+
+            # currently no-data value is still 255
+            tensor = torch.from_numpy(data).long()
+
+        elif modality in [
+            'aster',
+            'canopy_height_eth',
+            'sentinel1_asc',
+            'sentinel1_desc',
+            'sentinel2',
+            'era5',
+            'lat',
+            'lon',
+            'month',
+        ]:
+            data = data.astype(np.float32)
+            # See https://github.com/vishalned/MMEarth-train/blob/8d6114e8e3ccb5ca5d98858e742dac24350b64fd/mmearth_dataset.py#L88
+            # the modality is called sentinel2 but has different bands stats for l1c and l2a
+            if modality == 'sentinel2':
+                modality_ = (
+                    'sentinel2_l2a'
+                    if tile_info['S2_type'] == 'l2a'
+                    else 'sentinel2_l1c'
+                )
+            else:
+                modality_ = modality
+            data = self._normalize_modality(data, modality_, bands)
+            data = np.where(data == self.no_data_vals[modality], np.nan, data)
+            tensor = torch.from_numpy(data).float()
+        elif modality in ['biome', 'eco_region']:
+            data = data.astype(np.int32)
+            # no data value also 255 for biome and 65535 for eco_region
+            tensor = torch.from_numpy(data).long()
+        elif modality in [
+            'sentinel2_cloudmask',
+            'sentinel2_cloudprod',
+            'sentinel2_scl',
+        ]:
+            tensor = torch.from_numpy(data.astype(np.int32)).long()
+
+        # TODO: tensor might still contain nans, how to handle this?
+        return tensor
+
+    def _normalize_modality(
```

Comment:
I think this should be a datamodule transform, not hard-coded into the dataset.

Reply:
I think I would disagree with you here. I know that for most datasets, the normalization in torchgeo is only applied in the datamodule. I am also not a big fan of that (I think normalization and augmentation should be handled separately). Thinking ahead of training over multiple sources, normalization is unique to each data source, while things like geometric augmentations (flips etc) are applicable to all inputs. A datamodule then would have to involve all this normalization logic but for multiple cases depending on the data source, so to me this makes more sense.

Reply:
My counterarguments:

* TorchGeo is based on torchvision, which does not do normalization in the dataset, although you can specify transforms you want to apply
* Lightning intends its datamodules to be used for [splitting, transforming, normalizing, and tokenizing](https://lightning.ai/docs/pytorch/stable/data/datamodule.html#why-do-i-need-a-datamodule) the data
* Yes, normalizations are dataset-specific, but so are datamodules. We don't need generic datamodules that can be used for any dataset

I agree that multimodal datasets throw a wrench in this design, since each image needs a different normalization. I don't have a good counterargument for that, it is kind of annoying that we have to write custom transforms to handle each mode separately. I also fully agree that normalizations should be parameters so that I can specify them in my YAML file, although Lightning doesn't have a way to checkpoint these.

I don't disagree with you enough to hold up this PR, but I would like to decide on a consistent strategy for all datasets. Right now, our datasets are hundreds of lines of code, while our datamodules are only tens of lines of code. I want to try to shift as much of the complexity as possible to the datamodules if we can (for example, converting mask classes to ordinal numbers) so that the data loaders can focus on data loading and all transforms can be applied on the GPU.

Reply:
Yeah I am happy to discuss this at a later stage, just as a brief answer:

- torchvision expects the normalization on the dataset level. if you take any common dataset like imagenet, mnist, cifar etc,  everyone always has to copy that same transform snippet and normalization values from some other repo and pass it to the dataset class, which is redudant and seems unnecessary
- in the documentation lightning page in their example, the normalization is defined in a `DataModule` but is actually applied inside the `Dataset` through a transform function, no `after_batch_transform` function

Thus I don't think even lightning is fully clear on this, but at the top of my head I also don't have the perfect solution for torchgeo.

Reply:
Im happy to move the normalization function to wrappers to keep this streamlined

Reply:
@adamjstewart would you prefer, if I write a datamodule where all the normalization logic lives?

Reply:
I'll let you decide. I'm okay with both. I'm just trying to find ways to keep the datasets simple.

Reply:
Fair! I think I would keep it for now then, but open to changing it at some other point.